### ISSUE TITLE
`reportOverlappingOverload` False positive

### ISSUE THREAD

**Issue by shivaal-mako** [microsoft/pyright#9626]
**Describe the bug**
Pyright erroneously returns `Overload 1 for "test" overlaps overload 2 and returns an incompatible type (reportOverlappingOverload)` whereas Pylance does the right thing. I've seen this issue for a while now (couple months), only reporting it now.

**Code or Screenshots**
Problematic:
```python
from typing import Literal, overload

@overload
def test(a: int, b: int = 0, c: Literal[False] = False) -> int:
    ...

@overload
def test(a: int, b: int = 0, c: Literal[True] = True) -> str:
    ...

def test(a: int, b: int = 0, c: bool = True) -> int | str:
    if c:
        return f'{a+b}'
    return a+b
```

Working as intended:
```python
from typing import overload

@overload
def test2(a: int) -> int:
    ...

@overload
def test2(a: float) -> float:
    ...

def test2(a: int | float) -> int | float:
    return a
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?
`pyright 1.1.391` as command-line tool. Pylance inside VS Code does not give any errors.

**Comment by erictraut** [comment]
Pylance is built upon pyright, so their behaviors are the same unless you configure them differently. I suspect that you are either using an outdated version of pyright or pylance to do your comparison or you are using different configurations. Note that pylance defaults the `typeCheckingMode` to "off" whereas pyright defaults it to "standard".

Pyright's behavior in this case is correct, so this isn't a bug. You've defined two overloads that accept the same set of arguments but return incompatible (non-overlapping) return types. In particular, if you call `test(1)`, this matches both overloads, but the first overload indicates that the function returns an `int` and the second overload indicates that it returns a `str`. Only one of these can be correct.

To eliminate the error, you need to modify your second overload definition. 

```python
@overload
def test(a: int, b: int = 0, c: Literal[False] = False) -> int:
    ...

@overload
def test(a: int, b: int, c: Literal[True]) -> str:
    ...
```

**Comment by lukas-lang** [comment]
Ran into a similar issue recently, where overlapping overloads where detected after the code previously type-checked fine. I understand the logic behind why the overloads are technically overlapping, but it seems to me the new, stricter, enforcement makes it impossible to properly type hint this?
The fix suggested in the previous post doesn't really cover all the cases: What about `test(1, c= True)`? This should work at runtime, but none of the overloads cover it (since the default for `b` was removed from the second one). And I am not sure how I can type hint this at all now?

Side note: It seems the types are still inferred correctly despite the warning, so simply adding `# pyright: ignore[reportOverlappingOverload]` gets rid of any warnings but doesn't impact actual type inference - is this just a coincidence or is this reliable?

Ideally, one would be able to say
```python
def test(a: int, b: int = ..., c: Literal[True]) -> str: ...
```
but of course that doesn't work because of the non-optional parameter following `b`. This could simply be allowed in overloads (with the implication that the overload is only considered if the parameter is actually specified, either by name or by explicitly giving everything before), but I guess this would require `@overload` to be somewhat "magic" in terms of allowing otherwise unsupported syntax?
Alternatively, I could image a magic default value, e.g.
```python
def test(a: int, b: int = ..., c: Literal[True] = Specified) -> str: ...
```
This would work with existing syntax etc, while still hinting to the type checker that the overload should only be considered if the parameter is actually given.
